cloudinit/net/dhcp.py

```diff
@@ -267,101 +132,366 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
+            LOG.error(msg)
 
-        static_routes.append(("%s/%s" % (net_address, net_length), gateway))
+        current_idx = 0
+        for idx, tok in enumerate(tokens):
+            if idx < current_idx:
+                continue
+            net_length = int(tok)
+            if net_length in range(25, 33):
+                req_toks = 9
+                if len(tokens[idx:]) < req_toks:
+                    _trunc_error(net_length, req_toks, len(tokens[idx:]))
+                    return static_routes
+                net_address = ".".join(tokens[idx + 1 : idx + 5])
+                gateway = ".".join(tokens[idx + 5 : idx + req_toks])
+                current_idx = idx + req_toks
+            elif net_length in range(17, 25):
+                req_toks = 8
+                if len(tokens[idx:]) < req_toks:
+                    _trunc_error(net_length, req_toks, len(tokens[idx:]))
+                    return static_routes
+                net_address = ".".join(tokens[idx + 1 : idx + 4] + ["0"])
+                gateway = ".".join(tokens[idx + 4 : idx + req_toks])
+                current_idx = idx + req_toks
+            elif net_length in range(9, 17):
+                req_toks = 7
+                if len(tokens[idx:]) < req_toks:
+                    _trunc_error(net_length, req_toks, len(tokens[idx:]))
+                    return static_routes
+                net_address = ".".join(tokens[idx + 1 : idx + 3] + ["0", "0"])
+                gateway = ".".join(tokens[idx + 3 : idx + req_toks])
+                current_idx = idx + req_toks
+            elif net_length in range(1, 9):
+                req_toks = 6
+                if len(tokens[idx:]) < req_toks:
+                    _trunc_error(net_length, req_toks, len(tokens[idx:]))
+                    return static_routes
+                net_address = ".".join(
+                    tokens[idx + 1 : idx + 2] + ["0", "0", "0"]
+                )
+                gateway = ".".join(tokens[idx + 2 : idx + req_toks])
+                current_idx = idx + req_toks
+            elif net_length == 0:
+                req_toks = 5
+                if len(tokens[idx:]) < req_toks:
+                    _trunc_error(net_length, req_toks, len(tokens[idx:]))
+                    return static_routes
+                net_address = "0.0.0.0"
+                gateway = ".".join(tokens[idx + 1 : idx + req_toks])
+                current_idx = idx + req_toks
+            else:
+                LOG.error(
+                    'Parsed invalid net length "%s".  Verify DHCP '
+                    "rfc3442-classless-static-routes value.",
+                    net_length,
+                )
+                return static_routes
 
-    return static_routes
+            static_routes.append(
+                ("%s/%s" % (net_address, net_length), gateway)
+            )
 
+        return static_routes
+
+    @staticmethod
+    def get_dhclient_d():
+        # find lease files directory
```

Comment:
Good question. This is basically a direct copy/paste from DataSourceCloudStack, but modified to be in a class.

From the git history I see that:

- [6 years ago](https://github.com/canonical/cloud-init/commit/33816e96d8981918f734dab3ee1a967bce85451a) the network manager directory was added. I suspect that this is no longer required, since recently when digging around in upstream networkmanager code I recall seeing that isc-dhclient was dropped in favor of an internal implementation. A hasty grep of my local (ubuntu) NetworkManager-based system sees nothing that matches `dhcp-server-identifier`, which is the whole point of this function.
- [12 years ago](https://github.com/canonical/cloud-init/commit/4fde399a38765fa9641b3177b966ad6c8ec9750f) Ubuntu support was added, Fedora was already supported

I'm a bit hesitant to split this up, however, because of the risk of accidentally breaking one distro or another by making this more exact. Unless someone wants to audit all of the distros, I'd be more comfortable leaving it as is for now.